Added Support for Person Detection for ESP32

AdityaHPatwardhan opened the pull request:
Req- Labels : **comp: lite** , **Work In Progress**
Fix Some Build Issues
- [x]  1) remove  the file "person_model_grayscale/person_detect_model_data.cc” from CMakeLists.txt of main and add it in CMakeLists.txt of components/tfmicro/

- [x] 2) change the header person_detect_model_data.h in file  "person_model_grayscale/person_detect_model_data.cc” to its appropriate name

- [x] 3) Add support in current build system to add esp specific files to espressif projects such as sdkconfig.defaults , Kconfig.projbuilds, partitions.csv( if required) .

- [x] 4) Add esp specific third party components when building esp project for e.x. [esp32-camera](https://github.com/espressif/esp32-camera) for this project

To build this project successfully follow below steps till build issues are fixed 

1) Generate esp project using 
`make -f tensorflow/lite/experimental/micro/tools/make/Makefile TARGET=esp generate_person_detection_esp_project`

2) In esp-idf/main/CMakeLists.txt remove following line from SRCS var of idf_component_register `tensorflow/lite/experimental/micro/tools/make/downloads/person_model_grayscale/person_detect_model_data.cc` and add it to the SRCS var of comopnents/tfmicro/CMakeLists.txt  

3) in `person_detect_model_data.cc` (which is located at path given in step 2) change the included header file `person_detect_model_data` to its respective path.

4) Clone component  [esp32-camera](https://github.com/espressif/esp32-camera) into components/
5) Copy ` sdkconfig.defaults` from `person_detection/esp` and put it into newly generated `esp-idf` folder ( where we will build the project ) , also copy `Kconfig.projbuild` file from `person_detection/esp` into `esp-idf/main`
6) the idf at `$IDF_PATH`  must be checked out at release/v4.0

AdityaHPatwardhan committed:
Added Support for Person Detection for ESP32

AdityaHPatwardhan commented:
@fredrec , the `CMakelists.txt.tpl` had used wrong method to add compile flags so flag `-O3` ( along with some other flags) was not being applied while compiling, I have made the respective changes, Please test this example on ESP-EYE , I will add respective changes to the micro speech MR as well.

gbaned commented:
@AdityaHPatwardhan Could you please resolve the conflicts? Thanks!

AdityaHPatwardhan commented:
@gbaned @fredrec , did you get a chance to review this MR , It provides correct output within 1 sec of time, Please provide the patch which fixes the build issues. also If there are no issues pending please help merge it ASAP, Thank You

fredrec commented:
Hi Aditya,

I sent you a PR (https://github.com/AdityaHPatwardhan/tensorflow/pull/2) that should fix the build issues.

After generating the example (`make V=1 -f tensorflow/lite/micro/tools/make/Makefile TARGET=esp clean generate_person_detection_esp_project`), I was able to build by just adding the `esp32-camera` component and calling `idf.py`:
```
git clone https://github.com/espressif/esp32-camera.git components/esp32-camera
idf.py build
```

fredrec commented on tensorflow/lite/micro/examples/person_detection/README.md:
```diff
@@ -171,6 +171,69 @@ From the log, we can see that it took around 170 ms to capture and read the
+
+  * The `IDF_PATH` environment variable is set
+  * `idf.py` and Xtensa-esp32 tools (e.g. `xtensa-esp32-elf-gcc`) are in `$PATH`
+
```
It should also be mentioned that the `esp32-camera` component is needed.

BTW, would you recomend to clone it in `$IDF_PATH/components/` or in the `components/` dir of the example ?

AdityaHPatwardhan replied:
Yes I will update that in the `README` , I would recommend to clone it in `components/` dir of the example  (as we advice not to make any changes in `esp-idf` and keep it intact)

fredrec commented on tensorflow/lite/micro/examples/person_detection/esp/app_camera_esp.h:
```diff
@@ -0,0 +1,183 @@
+#define HREF_GPIO_NUM    CONFIG_CAMERA_PIN_HREF
+#define PCLK_GPIO_NUM    CONFIG_CAMERA_PIN_PCLK
+#endif
+
```
In the case when no known camera model is defined (I believe this can happen if the user forgot to configure), no obvious message will be shown, only some missing symbols errors.

Can you please add an `#else` clause at the end with an error message using the `#error` macro ? (Something like `#error Camera model is not configured`).

AdityaHPatwardhan replied:
@fredrec thank you for reviewing, Actually the default module is WROVER_MODULE, so it will be always selected in `Kconfig`. but anyways, I will add the necessary changes to make it foolproof.

fredrec reviewed (changes requested):
Thanks for your patience on this PR, only some minor fixes are needed to merge it.

fredrec committed:
Fix build for ESP person detector example (#2)

AdityaHPatwardhan commented:
@fredrec I have made the necessary changes, Thank you for the build fixes, and I think this should be ready to merge.

fredrec commented:
> @fredrec I have made the necessary changes, Thank you for the build fixes, and I think this should be ready to merge.

I just noticed that I mistakenly hard-coded some path in the make file. Can you please apply this fix: https://github.com/fredrec/tensorflow/commit/0cfd78254f36d3825e4164fd27aa0144fe762691

fredrec commented:
@petewarden: This PR seems to be done, it builds and works fine on an ESP-EYE board. Can you please merge it ?

AdityaHPatwardhan commented:
@fredrec I have made the suggested changes to build fixes. Thank you for your help in fixing the build issues.

AdityaHPatwardhan commented:
I fixed the compile warnings, when building the example, It runs successfully without any error.

fredrec commented:
Hi @AdityaHPatwardhan,

There are some problems merging this PR. Most likely it is because some files (`app_camera_esp.c`, `app_camera_esp.h`) have the wrong licence headers (currently `ESPRESSIF MIT License`). Can you please update it to the Tensorflow licence header ?

AdityaHPatwardhan commented:
@fredrec , Yes will do it ASAP.

AdityaHPatwardhan committed:
person_detection_esp: Fixed License for some files in person_detection/esp

AdityaHPatwardhan committed:
micro_speech_esp: Fixed build error for TARGET=esp in micro_speech example

AdityaHPatwardhan committed:
1) person_detection_esp: Fix README and
port build change from micro_s to person_d example
2) micro_speech_esp: Fix README

AdityaHPatwardhan commented:
@fredrec, I have made the respective changes regarding license headers as you requested, Along with that I have also Fixed README for the person_detection and micro_speech example.

Also,
When @tensorflower-gardener merged the micro_speech PR , I think they added a line 
``` ifeq ($(TARGET, “esp”)```  which causes build error in build micro_speech example on master branch 
which should have been ```ifeq ($(TARGET, esp) ```, I have fixed that error in this branch with 
[this-commit](https://github.com/tensorflow/tensorflow/commit/8978def1b7e6870edfc4dd20149e2c9e32b97dc6)
Along with this, 
There is a new error message  printed at the `generate_esp_project` command because of `ESP_PROJECT_FILES` which we added in the build fix of micro_speech.
When `generateing_esp_project` tensorflow is executing all the makefiles and internally the `ESP_PROJECT_FILES` is getting replaced with values for other examples which prints an error message while building a particular example (say hello_world)
Actually project is generated correctly , it just that `generate_esp_project` command prints failure.
Can you Please Take a look,
Thank You

rocLv commented:
[[@AdityaHPatwardhan I used exmail.qq.com to send email. I changed pem and builded seccessfully. But I can't receive emails.

The LOG was:
```
person detected
I (1212591) smtp_client: Connecting to smtp.exmail.qq.com:465...
I (1212641) smtp_client: Connected.
E (1272711) smtp_client: mbedtls_net_recv failed with error -0x0

I (1272711) smtp_client: Email sent successfully !
```

AdityaHPatwardhan commented:
@rocLv, if you are talking about the [doorbell_camera](https://github.com/espressif/tensorflow/tree/master/tensorflow/lite/micro/examples/doorbell_camera) example then I dont think this would be a proper platform to discuss the issue. As the example you are referring is in a different repository ( surely forked from this one, but still a different one ). If you have any issue regarding the `doorbell_camera` you may email me and we will take it up from there.
My email id - aditya.patwardhan@espressif.com
Thank you.
